### PR TITLE
18CZ Fixed Buying of trains

### PR DESCRIPTION
**Pull request by Zwergenpunk**
Fixed Buying of trains.

Corporations can buy other trains but have to buy at least one runnable train

### CODE REVIEW

File: lib/engine/game/g_18_cz.rb [hunk 1]
```diff
@@ -164,6 +164,26 @@ def block_lay_for_purple_tiles
...
+
+      def must_buy_train?(entity)
+        !entity.rusted_self &&
+        !depot.depot_trains.empty? &&
+        (entity.trains.empty? ||
+          (entity.type == :medium && entity.trains.none? { |item| !!(item.name =~ medium_train_regex) }) ||
+          (entity.type == :large && entity.trains.none? { |item| !!(item.name =~ large_train_regex) }))
+      end
+
+      def small_train_regex
```
Comment (tobymao) [inline lib/engine/game/g_18_cz.rb:176]: let's not do regexes and instead have a dictionary

File: lib/engine/step/g_18_cz/buy_train.rb [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+require_relative '../buy_train'
+
+module Engine
+  module Step
+    module G18CZ
+      class BuyTrain < BuyTrain
+        def buyable_train_variants(train, entity)
+          trains = super
+
+          return [] unless trains.any?
```
Comment (tobymao) [inline lib/engine/step/g_18_cz/buy_train.rb:12]: if trains.empty?

File: lib/engine/step/g_18_cz/buy_train.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+          default_trains = trains.select { |item| (item[:name] =~ @game.small_train_regex) }
```
Comment (tobymao) [inline lib/engine/step/g_18_cz/buy_train.rb:14]: probably have a helper fur this in game and again, don't use regexes

File: lib/engine/step/g_18_cz/buy_train.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+          return default_trains if entity.type == :small
+
+          medium_trains = trains.select { |item| (item[:name] =~ @game.medium_train_regex) }
+          if entity.type == :medium
+            return medium_trains if entity.trains.none? do |item|
+                                      (item.name =~ @game.medium_train_regex)
+                                    end && room_for_only_one?(entity)
+
+            return default_trains + medium_trains
```
Comment (tobymao) [inline lib/engine/step/g_18_cz/buy_train.rb:23]: use concat instead of + everywhere

File: lib/engine/game/g_18_cz.rb (hunk 1 above)
Comment (Zwergenpunk) [reply lib/engine/game/g_18_cz.rb:176]: yeah, thats maybe better ;)

File: lib/engine/game/g_18_cz.rb [hunk 5]
```diff
@@ -167,6 +173,24 @@ def block_lay_for_purple_tiles
...
+      end
+
+      def train_of_size?(item, size)
+        name = if item.is_a?(Hash)
+                 item[:name]
+               else
+                 item.name
+               end
+
+        TRAINS_FOR_CORPORATIONS[size].include?(name)
```
Comment (tobymao) [inline lib/engine/game/g_18_cz.rb:192]: wouldn't it be better to invert this? 

TRAINS_FOR_CORPORATIONS[name] == size

Comment (Zwergenpunk) [reply lib/engine/game/g_18_cz.rb:192]: Maybe? I think the dictionary is easy to read with coropration size => trains but possible is to have a list of trains => corporation size.

I can change it but I don't really think one is much better than the other one.

Comment (tobymao) [reply lib/engine/game/g_18_cz.rb:192]: let's change it

**Review by tobymao — APPROVED** [review]

**Pull request merged**